lib/src/path_parser.cc

```diff
@@ -0,0 +1,244 @@
+                    // separators, because we basically want "foo 3.0bar" to parse as a string even
+                    // though there's a number in it. The fact that we tokenize non-string values
+                    // is largely an implementation detail.
+                    auto value = tokens::get_value(t);
+
+                    // We need to split the tokens on a . so that we can get sub-paths but still preserve
+                    // the original path text when doing an insertion
+                    if (path_tokens != nullptr) {
+                        path_tokens->pop_back();
+                        token_list split_tokens = split_token_on_period(t, flavor);
+                        path_tokens->insert(path_tokens->end(), split_tokens.begin(), split_tokens.end());
+                    }
+                    text = value->transform_to_string();
+                } else if (t->get_token_type() == token_type::UNQUOTED_TEXT) {
+                    // We need to split the tokens on a . so that we can get sub-paths but still preserve
+                    // the original path text when doing an insertion on ConfigNodeObjects
+                    if (path_tokens != nullptr) {
+                        path_tokens->pop_back();
+                        token_list split_tokens = split_token_on_period(t, flavor);
+                        path_tokens->insert(path_tokens->end(), split_tokens.begin(), split_tokens.end());
+                    }
+                    text = t->token_text();
+                } else {
+                    throw config_exception("Invalid path: " + original_text +
+                                           " -- Token not allowed in path expression: " +
+                                           t->to_string() +
+                                           " (you can double-quote this token if you really want it here)");
+                }
+
+                add_path_text(elements, false, text);
+            }
+        }
+
+        path_builder builder;
+        for (element e : elements) {
+            if (e._value.length() == 0 && !e._can_be_empty) {
+                throw config_exception("Invalid path: " + original_text +
+                                               " -- path has a leading, trailing, or two adjacent '.'" +
+                " (use quoted \"\" empty string if you want an empty element)");
+            } else {
+                builder.append_key(e._value);
+            }
+        }
+
+        return builder.result();
+    }
+
+    token_list path_parser::split_token_on_period(shared_token t, config_syntax flavor) {
+        string token_text = t->token_text();
+        if (token_text == ".") {
+            return token_list { t };
+        }
+
+        auto token_it = boost::make_iterator_range(make_split_iterator(
+                token_text, token_finder(is_any_of("."), token_compress_on)), split_iterator<string::iterator>());
```

Comment:
The error seems to be between the return value of make_split_iterator() and the last argument. Either make_split_iterator() has to be made to return a const_iterator somehow, or we can't use const.